In [156]:
import pandas as pd
import numpy as np
import datetime

from matplotlib import font_manager, rc, rcParams
import matplotlib.pyplot as plt

In [9]:
import warnings
warnings.filterwarnings('ignore')

# 맵핑한 데이터 임포팅

In [32]:
df = pd.read_csv('news_stock_OK.csv', encoding = 'utf-8 sig')
df.head()

Unnamed: 0        Date                                              NOUNS  \
0          99  2017-12-15  ['부고', '송창', '씨', '법무', '법인', '세한', '변호사', '부친...   
1         102  2017-12-18  ['주가', '반토막', '삼성', '중공업', '뒷북', '증권사', '뉴스', ...   
2         103  2017-12-19  ['배당', '계절', '개인', '접근', '로', '배당', '주', '투자',...   
3         105  2017-12-21  ['공정위', '삼성', '물산', '합병', '가이드라인', '개정', '논란',...   
4         106  2017-12-22  ['스팟', '삼성', '전자', '오른', '출발', '삼성', '전자', '출발...   

      Open     High      Low    Close     Adj Close      Volume    day1C  \
0  51240.0  51480.0  50520.0  50620.0  17585.119141  14928550.0  51060.0   
1  50620.0  51240.0  50620.0  51200.0  17786.607422   7350250.0  50620.0   
2  51540.0  52080.0  51520.0  51560.0  17911.669922  11978600.0  51200.0   
3  51000.0  51060.0  49100.0  49140.0  17070.974609  15624300.0  51560.0   
4  49400.0  49960.0  49240.0  49700.0  17265.515625  11199650.0  49140.0   

      ...       Diff5  Diff10  Diff1_per  Diff3_per  Diff5_per  Diff10_per  \
0     ...     -1380.0  -220.0      -0.87      -2.92      -2.73       -0.43   
1     ...      -580.0  -140.0       1.13      -0.23      -1.13       -0.27   
2     ...      -540.0   300.0       0.70       0.97      -1.05        0.58   
3     ...     -2180.0  -880.0      -4.92      -3.01      -4.44       -1.79   
4     ...     -1360.0 -1040.0       1.13      -3.02      -2.74       -2.09   

   Diff1_clf  Diff3_clf  Diff5_clf  Diff10_clf  
0       -1등급       -2등급       -2등급        -1등급  
1       +2등급       -1등급       -2등급        -1등급  
2       +1등급       +1등급       -2등급        +1등급  
3       -3등급       -3등급       -3등급        -2등급  
4       +2등급       -3등급       -2등급        -2등급  

[5 rows x 37 columns]

### 칼럼 너무 많아서 df_original에다가 보존시키고 내가 쓰는 것만 df로 작업

In [11]:
# 복제
df_original = df.copy()
df.columns

Index(['Unnamed: 0', 'Date', 'NOUNS', 'Open', 'High', 'Low', 'Close',
       'Adj Close', 'Volume', 'day1C', 'day3C', 'day5C', 'day10C', 'day1V',
       'day3V', 'day5V', 'day10V', 'ma1C', 'ma3C', 'ma5C', 'ma10C', 'ma1V',
       'ma3V', 'ma5V', 'ma10V', 'Diff1', 'Diff3', 'Diff5', 'Diff10',
       'Diff1_per', 'Diff3_per', 'Diff5_per', 'Diff10_per', 'Diff1_clf',
       'Diff3_clf', 'Diff5_clf', 'Diff10_clf'],
      dtype='object')

In [18]:
# 칼럼 삭제
df.drop(['Unnamed: 0', 'High', 'Low', 'Adj Close', 'Volume',
       'day1C', 'day3C', 'day5C', 'day10C', 'day1V', 'day3V', 'day5V',
       'day10V', 'ma1C', 'ma3C', 'ma5C', 'ma10C', 'ma1V', 'ma3V', 'ma5V',
       'ma10V'], axis=1, inplace = True)

In [13]:
df.columns

Index(['Date', 'NOUNS', 'Open', 'Close', 'Diff1', 'Diff3', 'Diff5', 'Diff10',
       'Diff1_per', 'Diff3_per', 'Diff5_per', 'Diff10_per', 'Diff1_clf',
       'Diff3_clf', 'Diff5_clf', 'Diff10_clf'],
      dtype='object')

In [14]:
df.head()

Date                                              NOUNS     Open  \
0  2017-12-15  ['부고', '송창', '씨', '법무', '법인', '세한', '변호사', '부친...  51240.0   
1  2017-12-18  ['주가', '반토막', '삼성', '중공업', '뒷북', '증권사', '뉴스', ...  50620.0   
2  2017-12-19  ['배당', '계절', '개인', '접근', '로', '배당', '주', '투자',...  51540.0   
3  2017-12-21  ['공정위', '삼성', '물산', '합병', '가이드라인', '개정', '논란',...  51000.0   
4  2017-12-22  ['스팟', '삼성', '전자', '오른', '출발', '삼성', '전자', '출발...  49400.0   

     Close   Diff1   Diff3   Diff5  Diff10  Diff1_per  Diff3_per  Diff5_per  \
0  50620.0  -440.0 -1480.0 -1380.0  -220.0      -0.87      -2.92      -2.73   
1  51200.0   580.0  -120.0  -580.0  -140.0       1.13      -0.23      -1.13   
2  51560.0   360.0   500.0  -540.0   300.0       0.70       0.97      -1.05   
3  49140.0 -2420.0 -1480.0 -2180.0  -880.0      -4.92      -3.01      -4.44   
4  49700.0   560.0 -1500.0 -1360.0 -1040.0       1.13      -3.02      -2.74   

   Diff10_per Diff1_clf Diff3_clf Diff5_clf Diff10_clf  
0       -0.43      -1등급      -2등급      -2등급       -1등급  
1       -0.27      +2등급      -1등급      -2등급       -1등급  
2        0.58      +1등급      +1등급      -2등급       +1등급  
3       -1.79      -3등급      -3등급      -3등급       -2등급  
4       -2.09      +2등급      -3등급      -2등급       -2등급

# 이제 통계분석 시작
- https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
- https://ratsgo.github.io/machine%20learning/2017/05/18/naive/
1. Naive Basian

In [71]:
df.head()

Unnamed: 0        Date                                              NOUNS  \
0          99  2017-12-15  ['부고', '송창', '씨', '법무', '법인', '세한', '변호사', '부친...   
1         102  2017-12-18  ['주가', '반토막', '삼성', '중공업', '뒷북', '증권사', '뉴스', ...   
2         103  2017-12-19  ['배당', '계절', '개인', '접근', '로', '배당', '주', '투자',...   
3         105  2017-12-21  ['공정위', '삼성', '물산', '합병', '가이드라인', '개정', '논란',...   
4         106  2017-12-22  ['스팟', '삼성', '전자', '오른', '출발', '삼성', '전자', '출발...   

      Open     High      Low    Close     Adj Close      Volume    day1C  \
0  51240.0  51480.0  50520.0  50620.0  17585.119141  14928550.0  51060.0   
1  50620.0  51240.0  50620.0  51200.0  17786.607422   7350250.0  50620.0   
2  51540.0  52080.0  51520.0  51560.0  17911.669922  11978600.0  51200.0   
3  51000.0  51060.0  49100.0  49140.0  17070.974609  15624300.0  51560.0   
4  49400.0  49960.0  49240.0  49700.0  17265.515625  11199650.0  49140.0   

      ...       Diff5  Diff10  Diff1_per  Diff3_per  Diff5_per  Diff10_per  \
0     ...     -1380.0  -220.0      -0.87      -2.92      -2.73       -0.43   
1     ...      -580.0  -140.0       1.13      -0.23      -1.13       -0.27   
2     ...      -540.0   300.0       0.70       0.97      -1.05        0.58   
3     ...     -2180.0  -880.0      -4.92      -3.01      -4.44       -1.79   
4     ...     -1360.0 -1040.0       1.13      -3.02      -2.74       -2.09   

   Diff1_clf  Diff3_clf  Diff5_clf  Diff10_clf  
0       -1등급       -2등급       -2등급        -1등급  
1       +2등급       -1등급       -2등급        -1등급  
2       +1등급       +1등급       -2등급        +1등급  
3       -3등급       -3등급       -3등급        -2등급  
4       +2등급       -3등급       -2등급        -2등급  

[5 rows x 37 columns]

In [20]:
# train_test 나누기
from sklearn.model_selection import train_test_split

# X에는 뉴스에서 추출한 명사 데이터만
X = df['NOUNS']

# y에는 주가상승/하락 정리한 범주형 데이터로 비교
y = df.loc[:, ['Diff1_clf', 'Diff3_clf', 'Diff5_clf', 'Diff10_clf']]

# X훈련,테스트 / y훈련,테스트
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

## 분석 전, 데이터 가공 (토큰화, 벡터 -> tf/idf -> 분석 -> 테스트)
### Tokenizing text with
- https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html#tokenizing-text-with-scikit-learn
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
- https://datascienceschool.net/view-notebook/3e7aadbf88ed4f0d87a76f9ddc925d69/

In [33]:
# 데이터 토크나이징 전
print(X_train.shape)
print(X_test.shape)

print(y_train.shape)
print(y_test.shape)

(194,)
(49,)
(194, 4)
(49, 4)


In [24]:
# 데이터 벡터 형태로 토크나이징
# 이 방법으로 토크나이징 하는게 맞나?? tf/idf같은 형식도 있던데??
# 스탑워드는 따로 처리 안해도 되나??
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()

# fit과 fit_transform의 차이는?? return term-document matrix 추가진행
# fit_transform(..)은 두 과정을 한 방에 하려고 붙인 것
X_train_counts = count_vect.fit_transform(X_train)  
X_train_counts.shape

# fit은 shape로 모양을 찍어볼 수 없다.

(194, 27130)

In [25]:
# 이 함수는 특정 단어의 위치? 갯수? 를 가져오는듯
# counts of N-grams of words or consecutive characters

count_vect.vocabulary_.get(u'잘못')

19009

### 헷갈리니 중간점검
- (194,)형태의 X_train데이터를 토큰화? 벡터화?
- X_train_counts  라는 이름으로 fit과 transform진행

## 문서 길이가 길어지면 카운팅 값이 자연스레 높아져서 뭔가 불공평함
- 그럴 때 tf/idf 토크나이징 사용...이 맞나??
- 다른 단어와의 상대적 중요성이 필요할 때는 좋을 것 같지만, 주가 예측에 어떤 단어가 핫했나를 따지려면 굳이 필요 없을 듯...??

In [35]:
# 토크나이징한 x의 벡터값들을 tf로 변환하는 과정
# fit과 transform 진행
from sklearn.feature_extraction.text import TfidfTransformer 

# 아까 벡터화한거 넣어서 fit
tf_transformer = TfidfTransformer(use_idf=False).fit(X_train_counts) 
print(tf_transformer)

# fit한걸 transform진행
X_train_tf = tf_transformer.transform(X_train_counts) 
print(X_train_tf.shape)

TfidfTransformer(norm='l2', smooth_idf=True, sublinear_tf=False,
         use_idf=False)
(194, 27130)


### 위랑 비슷하지만 idf도 같이 한 버전??

In [36]:
tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
print(X_train_tfidf.shape)

(194, 27130)


In [72]:
# 학습시킬 때 칼럼이 두 개 있으면 오류가 나서 일자별로 자르는 과정
y_train1 = y_train.iloc[:,0]
y_train2 = y_train.iloc[:,1]
y_train3 = y_train.iloc[:,2]
y_train4 = y_train.iloc[:,3]

y_test1 = y_test.iloc[:,0]
y_test2 = y_test.iloc[:,1]
y_test3 = y_test.iloc[:,2]
y_test4 = y_test.iloc[:,3]

In [168]:
y_test1

47     -1등급
211    -1등급
101    +1등급
153    +2등급
179    -2등급
188    -1등급
167    +1등급
9      +2등급
151    +1등급
183    +2등급
4      +2등급
63     -3등급
131    -1등급
178    -2등급
206    +1등급
21     -2등급
7      +3등급
170    +2등급
160    -1등급
198    +2등급
119    -2등급
29     +1등급
91     -1등급
242    -1등급
192    +1등급
152    +1등급
240    -2등급
135    -1등급
79     +2등급
57     -1등급
93     +2등급
96     -1등급
59     -1등급
71     -2등급
116    -2등급
42     -1등급
121    -1등급
11     +2등급
177    -2등급
203    -2등급
74     +2등급
216    +2등급
117    -2등급
2      +1등급
88     +2등급
10     -2등급
120    +1등급
229    -1등급
89     -1등급
Name: Diff1_clf, dtype: object

In [42]:
print(X_train_counts.shape)

print(X_train_tf.shape)
print(X_train_tfidf.shape)

print(y_train1.shape)
print(y_test1.shape)

(194, 27130)
(194, 27130)
(194, 27130)
(194,)
(49,)


In [ ]:
x

# 나이브 베이지안
- 을 쓰려면 y가 범주형 데이터로 들어가야 하지 않나...?

In [83]:
from sklearn.naive_bayes import MultinomialNB
clf_tf = MultinomialNB().fit(X_train_tf, y_train1)
clf_tfidf = MultinomialNB().fit(X_train_tfidf, y_train1)    # X_train_tfidf, twenty_train.target

# 여기 예제에서 y값(twenty_train.target)은 n=4 카테고리컬 변수인데, 내가 필요한 주가 diff를 범주화해야하나??

In [84]:
print(clf_tf)
print(clf_tfidf)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)
MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)


## 학습시킨 모델을 Test셋으로 비교

In [89]:
# Test셋 벡터로 변환
X_test_counts = count_vect.transform(X_test)

# Test셋 벡터값을 tf랑  tfidf로 변환
X_train_tf = tf_transformer.transform(X_test_counts)
X_train_tfidf = tfidf_transformer.transform(X_test_counts)


predicted_tf = clf_tf.predict(X_train_tf)
predicted_tfidf = clf_tfidf.predict(X_train_tfidf)

'''
for doc, category in zip(X_train, predicted):
    print('%r => %s' % (doc, [category]))    
'''

# 'God is love' => soc.religion.christian
# 'OpenGL on the GPU is fast' => comp.graphics

"\nfor doc, category in zip(X_train, predicted):\n    print('%r => %s' % (doc, [category]))    \n"

In [90]:
# 예측한 값이 predicted에 들어감
print(predicted_tf.shape)  # X_train->벡터->tf를 학습시켜 분류한 결과
print(predicted_tfidf.shape)  # X_train->벡터->tfidf를 학습시켜 분류한 결과

# 비교에 쓸 y_test 형태 파악
print(y_test1.shape)

(49,)
(49,)
(49,)


In [162]:
print('======== 1일치 예측 ========')
print('tf로 변환한 예측률 :', np.mean(predicted_tf == y_test1))
print('tfidf로 변환한 예측률 :',np.mean(predicted_tfidf == y_test1))

======== 1일치 예측 ========
tf로 변환한 예측률 : 0.22448979591836735
tfidf로 변환한 예측률 : 0.22448979591836735


In [92]:
print('======== 3일치 예측 ========')
print('tf로 변환한 예측률 :', np.mean(predicted_tf == y_test2))
print('tfidf로 변환한 예측률 :',np.mean(predicted_tfidf == y_test2))

print('======== 5일치 예측 ========')
print('tf로 변환한 예측률 :', np.mean(predicted_tf == y_test3))
print('tfidf로 변환한 예측률 :',np.mean(predicted_tfidf == y_test3))

print('======== 10일치 예측 ========')
print('tf로 변환한 예측률 :', np.mean(predicted_tf == y_test4))
print('tfidf로 변환한 예측률 :',np.mean(predicted_tfidf == y_test4))

======== 3일치 예측 ========
tf로 변환한 예측률 : 0.30612244897959184
tfidf로 변환한 예측률 : 0.30612244897959184
======== 5일치 예측 ========
tf로 변환한 예측률 : 0.16326530612244897
tfidf로 변환한 예측률 : 0.16326530612244897
======== 10일치 예측 ========
tf로 변환한 예측률 : 0.14285714285714285
tfidf로 변환한 예측률 : 0.12244897959183673


# Building a pipeline
- 분석기 이름 : nb_clf

In [136]:
from sklearn.pipeline import Pipeline
nb_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

#  Evaluation of the performance on the 'Test' set
- 파이프라인에는 tfidf사용

In [137]:
# X트레인과 1일치y트레인 학습시켜 모델 구축
nb_clf.fit(X_train, y_train1) 

# 만든 모델로 X테스트 학습 => y테스트 예측률 평가
nb_predicted1 = nb_clf.predict(X_test)

print('======== TestSet 1일치 예측 ========')
np.mean(nb_predicted1 == y_test1) 

======== TestSet 1일치 예측 ========


0.22448979591836735

In [138]:
text_clf.fit(X_train, y_train2) 
nb_predicted2 = text_clf.predict(X_test)

print('======== TestSet 3일치 예측 ========')
np.mean(nb_predicted2 == y_test2) 

======== TestSet 3일치 예측 ========


0.1836734693877551

In [179]:
text_clf.fit(X_train, y_train3) 
nb_predicted3 = text_clf.predict(X_test)

print('======== TestSet 5일치 예측 ========')
np.mean(nb_predicted3 == y_test3) 

======== TestSet 5일치 예측 ========


0.20408163265306123

In [140]:
text_clf.fit(X_train, y_train4) 
nb_predicted4 = text_clf.predict(X_test)

print('======== TestSet 10일치 예측 ========')
np.mean(nb_predicted4 == y_test4) 

======== TestSet 10일치 예측 ========


0.22448979591836735

## SVM

In [130]:
from sklearn.linear_model import SGDClassifier
svm_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier(loss='hinge', penalty='l2',
                          alpha=1e-3, random_state=42,
                          max_iter=5, tol=None)),
])

svm_clf.fit(X_train, y_train4)  

predicted = svm_clf.predict(X_test)
np.mean(predicted == y_test4)

0.22448979591836735

In [112]:
from sklearn import metrics
print(metrics.classification_report(y_test1, test_predicted1,
    target_names=y_test1))

             precision    recall  f1-score   support

       -1등급       0.00      0.00      0.00         9
       -1등급       0.00      0.00      0.00        12
       +1등급       0.00      0.00      0.00         1
       +2등급       0.33      0.07      0.11        15
       -2등급       0.23      0.91      0.36        11
       -1등급       0.00      0.00      0.00         1

avg / total       0.15      0.22      0.12        49



In [171]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test1, test_predicted1)


array([[ 0,  0,  0,  0,  9,  0],
       [ 0,  0,  0,  1, 11,  0],
       [ 0,  0,  0,  0,  1,  0],
       [ 2,  0,  0,  1, 12,  0],
       [ 0,  0,  0,  1, 10,  0],
       [ 0,  0,  0,  0,  1,  0]], dtype=int64)

In [175]:
y_test1.value_counts()

-1등급    15
+2등급    12
-2등급    11
+1등급     9
-3등급     1
+3등급     1
Name: Diff1_clf, dtype: int64

In [176]:
y_test2.value_counts()

-2등급    16
+1등급     7
+2등급     7
-3등급     7
+3등급     7
-1등급     4
-4등급     1
Name: Diff3_clf, dtype: int64

In [177]:
y_test3.value_counts()

-3등급    9
-2등급    8
+3등급    8
+2등급    8
-4등급    8
+4등급    3
-1등급    3
+1등급    2
Name: Diff5_clf, dtype: int64

In [178]:
y_test4.value_counts()

-4등급    14
-2등급     8
+2등급     6
+4등급     5
-3등급     5
+1등급     4
+3등급     4
-1등급     2
5등급      1
Name: Diff10_clf, dtype: int64

In [ ]:
precision = 예측한것 중에 맞힌 비율  => 예측을 안하면 올라가고 recall이 떨어짐 (주식은 이게 중요 / 오답이 치명적이라) => 일부러 피팅시키기도 함
recall = 전체 맞는 것 중 맞힌 비율  => reacall이 높으면 (검색은 이거)

# Parameter tuning using grid search

In [103]:
from sklearn.model_selection import GridSearchCV
parameters = {
    'vect__ngram_range': [(1, 1), (1, 2)],
    'tfidf__use_idf': (True, False),
    'clf__alpha': (1e-2, 1e-3),   # 0.01 or 0.001
}

In [104]:
gs_clf = GridSearchCV(text_clf, parameters, cv=5, iid=False, n_jobs=-1)

In [106]:
gs_clf = gs_clf.fit(X_train, y_train1)

In [ ]:
text_clf.predict(X_test)

In [108]:
Grid_predicted = gs_clf.predict(X_test)

In [110]:
# 하이퍼 파라미터 돌린 모델 예측률 추정
np.mean(Grid_predicted == y_test1) 

0.24489795918367346

In [109]:
gs_clf.best_score_                                  

for param_name in sorted(parameters.keys()):
    print("%s: %r" % (param_name, gs_clf.best_params_[param_name]))

clf__alpha: 0.01
tfidf__use_idf: True
vect__ngram_range: (1, 2)


In [160]:
%matplotlib inline

rcParams['axes.unicode_minus'] = False  # - 기호 깨짐 방지
font_fname = 'C:\Windows\Fonts\D2Coding.ttf'
font_family = font_manager.FontProperties(fname=font_fname).get_name()
print(font_family)
rc('font', family=font_family)

from matplotlib import font_manager
font_manager._rebuild()

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Windows\\Fonts\\D2Coding.ttf'

In [166]:
y_test

Diff1_clf Diff3_clf Diff5_clf Diff10_clf
47       -1등급      -1등급      -1등급       +2등급
211      -1등급      +3등급      +4등급       +1등급
101      +1등급      +1등급      -3등급       -4등급
153      +2등급      +2등급      +1등급       +2등급
179      -2등급      -3등급      -3등급       -4등급
188      -1등급      -2등급      +1등급       +3등급
167      +1등급      +1등급      +3등급       +3등급
9        +2등급      +3등급      +3등급       +2등급
151      +1등급      -2등급      -2등급       -3등급
183      +2등급      +1등급      +3등급       -1등급
4        +2등급      -3등급      -2등급       -2등급
63       -3등급      -2등급      -2등급       -1등급
131      -1등급      +1등급      -2등급       -2등급
178      -2등급      -1등급      -3등급       -4등급
206      +1등급      -2등급      -4등급       -4등급
21       -2등급      -2등급      +2등급       -4등급
7        +3등급      +2등급      -2등급       -2등급
170      +2등급      +3등급      +3등급       +4등급
160      -1등급      -2등급      -4등급       -3등급
198      +2등급      +1등급      -2등급       -3등급
119      -2등급      -4등급      -4등급       -4등급
29       +1등급      -2등급      +2등급       +1등급
91       -1등급      -1등급      +3등급       +3등급
242      -1등급      -2등급      -3등급       -4등급
192      +1등급      -2등급      -3등급       -2등급
152      +1등급      -2등급      -2등급       -2등급
240      -2등급      -3등급      -4등급       -4등급
135      -1등급      +2등급      -1등급       -3등급
79       +2등급      +2등급      +2등급       +3등급
57       -1등급      +3등급      +3등급       +4등급
93       +2등급      -1등급      -1등급       +2등급
96       -1등급      -2등급      -3등급       +2등급
59       -1등급      -2등급      +2등급        5등급
71       -2등급      -3등급      -3등급       -4등급
116      -2등급      -2등급      -3등급       -4등급
42       -1등급      -3등급      +3등급       -2등급
121      -1등급      -2등급      -4등급       -4등급
11       +2등급      +2등급      +4등급       +2등급
177      -2등급      -2등급      -4등급       -3등급
203      -2등급      -3등급      -3등급       -4등급
74       +2등급      +3등급      +2등급       -2등급
216      +2등급      +2등급      +3등급       +4등급
117      -2등급      -3등급      -4등급       -4등급
2        +1등급      +1등급      -2등급       +1등급
88       +2등급      +3등급      +2등급       +4등급
10       -2등급      +1등급      +4등급       +1등급
120      +1등급      -2등급      -4등급       -4등급
229      -1등급      +2등급      +2등급       -2등급
89       -1등급      +3등급      +2등급       +4등급

In [154]:
y_train4.unique()

array(['-4등급', '+2등급', '-2등급', '-5등급', '-1등급', '+4등급', '+3등급', '-3등급',
       '+1등급', '5등급'], dtype=object)

In [150]:
y_test4.unique()

array(['+2등급', '+1등급', '-4등급', '+3등급', '-3등급', '-1등급', '-2등급', '+4등급',
       '5등급'], dtype=object)

In [ ]:
fitting model on training data and transforming to test data????

In layman's terms, fit_transform means to do some calculation and then do transformation 
(say calculating the means of columns from some data and then replacing the missing values). 

So for training set, you need to both calculate and do transformation.

But for testing set, Machine learning applies prediction based on what was learned 
during the training set and so it doesn't need to calculate, it just performs the transformation.

In [ ]:
Code
import numpy as np
X_train, X_test, y_train, y_test = X[:60000], X[60000:], y[:60000], y[60000:]
shuffle_index = np.random.permutation(60000)
X_train, y_train = X_train[shuffle_index], y_train[shuffle_index]
# 훈련 세트를 섞어 모든 교차 검증 폴드가 비슷해지도록 만들었습니다. 특정 학습 알고리즘은 훈련 샘플의 순서에 민감하게 반응해서 많은 비슷한 샘플이 연이어 나타나면 성능이 나빠집니다. 데이터셋을 섞으면 이러한 문제를 방지할 수 있습니다.

In [ ]:
# shift()함수는 밑으로 내리는 것
# 그 전날꺼를 오른쪽에 붙이는 거 
# day1C는 1일 전꺼의 가격을 의미
df['day1C'] = df['Close'].shift(1)